### ISSUE TITLE
'tput reset' (cls-like command) won't work

### ISSUE THREAD

**Issue by japm48** [Microsoft/WSL#1456]
Related to: #1369 #1441 (and probably others)
* **Description**: I usually define `alias cls='tput reset'` to get a unix equivalent for 
* **Build number**: 14393.447 (but apparently still not solved).
* **Steps to reproduce**: just maximize the console window and execute `tput reset`
* **Expected result**: the full screen scrollback clears (like cls in cmd.exe).
* **Actual result**: interestingly, not only occurs what happens with just `clear` (like the other reports), but also the actual **console window shrinks** (smaller vertical size).

**Notes**:
What `tput reset` does is similar to the related bugs: it prints an escape code (obtained from terminfo database).
In particular, you can get the same result with: `echo -e '\033c\033[!p\033[?3;4l\033[4l\033>'`
This escape code was extracted by this little trick: `tput reset | sed -n l`

Trying with another `TERM` produces similar results: `TERM=rxvt tput reset`.

Additional info: https://www.gnu.org/software/termutils/manual/termutils-2.0/html_chapter/tput_1.html

**Comment by fpqc** [comment]
Isn't \033c not fully implemented?  There is another thread that.

**Comment by japm48** [comment]
Ah, you are right: I just checked it `tput reset` is sending `\033c\033[!p\033[?3;4l\033[4l\033>` and the part after `\033c` is the same as `tput init` (equivalent to `'printf \033[!p\033[?3;4l\033[4l\033>'`).
Apparently, I forgot that I also chose `tput reset` in case I messed with the terminal (ej. ncurses) or used cat on a binary file by accident. 
So, I guess, I restrict the scope of this report to the resizing behaviour (this it doesn't happen in linux).

**Comment by fpqc** [comment]
Does it set the terminal exactly back to 80x24 (size of a hardware terminal)?

**Comment by japm48** [comment]
I just counted: it sets width to 80 columns and doesn't modify the height.
Is this some kind of expected behaviour? Again, I never remember something similar in linux, but probably because of the particular terminal programs I used.

**Comment by fpqc** [comment]
@japm48 Yup. Look at the subsequence `033[?3;4l`.  This says "set 80 column mode and fast scroll mode" according to the [xterm documentation](http://www.xfree86.org/4.5.0/ctlseqs.html)

**Comment by fpqc** [comment]
So aside from the broken \033c, I have to say that this is the expected behavior according to the spec.

**Comment by japm48** [comment]
OK, I just tried with a Linux install, installed xterm and typed those commands, there is no resize behaviour at all. But according to the spec it does what is expected.
Thanks for the explanation.

**Comment by fpqc** [comment]
@japm48 On a Linux version of xterm, does tput print that control sequence?

**Comment by japm48** [comment]
I just made some tests.
`tput reset` prints:
 * In Xterm and Konsole (they set `TERM=xterm`):
`\033c\033[!p\033[?3;4l\033[4l\033>` (same as before)
 * In LXTerminal and Gnome Terminal (they set `TERM=xterm-256color`):
`\033c\033]104\a\033[!p\033[?3;4l\033[4l\033>`
 * In a ssh session (`TERM=rxvt`):
`\033>\033[1;3;4;5;6l\033[?7h\033[m\033[r\033[2J\033[H\033[r\033[m\033\*[2J\033[H\033[?7h\033[?1;3;4;6l\033[4l\033=\033[?1000l\033[?25h`
(there is "new line" in `*` I don't know how to escape that)

In the GUI window is never resized in those cases.

Also:
 * In a getty virtual console (`Ctrl+Alt+Fn`) (`TERM=linux`):
`\033c\033]R`

But of course there is no GUI

**Comment by fpqc** [comment]
@japm48 Yeah, I already figured it out. `xterm` ignores those control sequences unless it is launched with the mode `-132`.

**Comment by japm48** [comment]
OK, you just answered just after I pressed send :)
In that case, should the default behaviour be changed to match the Linux defaults?

**Comment by fpqc** [comment]
@japm48 I dunno, we'll see.  The console team probably just implemented that control sequence because they were reading down the control sequence documentation and just did it, not realizing that it was ignored on real `xterm`.

**Comment by japm48** [comment]
Thanks. It's interesting to learn all that, and I didn't know where to search.
I'll wait for an official response, meanwhile I'll be using mintty, which behaves more like linux.

**Comment by fpqc** [comment]
@japm48 You give me too much credit. I just read the terminfo/ncurses FAQ yesterday and then did some web-search-fu.

**Comment by zadjii-msft** [comment]
@japm48 @fpqc 
Thanks for shedding light on this. Yes, I implemented `CSI ? 3 h/l` correctly, initially because we were using `vttest` to judge our terminal capabilities, and that's what it uses to clear the screen and setup (IIRC). 

This is just one of many cases where the actual behavior of the terminal does not directly match any of the previous specs I read, but instead was documented in another previously unknown spec. I'll make a bug number for tracking and disabling this by default.

Yes, `ESC c` is not implemented at the time, I might be able to slip that in with fixing this though.

**Comment by fpqc** [comment]
@zadjii-msft You know, it would be pretty cool if you teamed up with the Powershell people and added something like "Powerterm", which would spawn a terminal-processing-mode-compatible version of Powershell in a conhost window with terminal processing enabled (They should be able to backport the code from the Linux edition of Powershell to the  Windows one).  

This would also fix that Win32-Openssh problem I mentioned over on that issues page (the one I pinged you on).

**Comment by jdebp** [comment]
The right equivalent for `CLS` is surely `tput clear` anyway.  That is, in effect, what the `clear_console` command does.

**Comment by therealkenc** [comment]
Closing dangling console tag. If I follow this one, it correctly it [came down to](https://github.com/Microsoft/WSL/issues/1456#issuecomment-264751998):

> so, I should restrict the scope of this issue to the window resizing behaviour (this it doesn't happen in linux).

I just tried `printf '\e[8;50;100t'` and the terminal resize sequences appear to work for me in Windows Console on 17713. What sequence `tput reset` sends is controlled by your `TERM` variable and terminfo. There's really no "correct" answer to that, and there is no specific "Linux" behaviour. [DEC VT200 terminals are notoriously difficult to resize without a trash compactor.] Further discussion to the [Windows Console Issue Tracker](https://github.com/microsoft/console).
